### PR TITLE
fix(downloader): resume large downloads instead of restarting from zero

### PR DESCRIPTION
**Pull request by caiopizzol**
Retries used to reopen the target file with truncation, so any download that failed mid-file restarted from byte zero - against a server that stalls long streams (as Receita's does under load), a large Empresas zip could never complete no matter how many retries were configured.

- Downloads stream into a .part file and every retry resumes from the bytes already on disk via a Range request
- A server that ignores Range gets a clean restart, never a corrupted concatenation; the final zip only appears via atomic rename once the size matches the expected total
- Finished files are zip-validated (CRC via testzip) before extraction, so truncation or corruption is a retryable download failure instead of a crash later
- A 416 response for an already-complete partial finalizes it instead of failing
- New tests cover mid-file stalls, offset resume, Range-ignoring servers, size mismatch, and corrupt archives

---
## Summary by cubic
Fixes the downloader so interrupted ZIP downloads resume instead of restarting from zero. Uses month‑scoped `.part` files with HTTP Range and `Accept-Encoding: identity`, CRC‑validates cached and new ZIPs, and preserves `.part` files across runs to safely resume.

- Bug Fixes
  - Resume logic: write to month‑scoped `.part`, send `Range` and `Accept-Encoding: identity`, validate `Content-Range`/`Content-Length`, clean‑restart when `Range` is ignored, rename to final only when sizes match, raise `DownloadIncompleteError` on mismatches; keep `.part` files across runs for cross‑run resume.
  - ZIP integrity: CRC‑verify downloads with `testzip` before extraction and validate cached ZIPs before reuse; delete bad cached ZIPs and re‑download; treat truncation, CRC corruption, and overlong bodies as retryable; never publish a bad `.zip`.
  - Edge cases and tests: finalize on 416 when totals match; discard partials on wrong offsets, mismatched or missing headers, unexpected statuses, or cross‑month/stale partials; add tests for 206 without `Content-Length`, malformed/reversed `Content-Range`, 416 with/without totals, preexisting final ZIP replacement, corrupt archives, and corrupt cached ZIPs.

<sup>Written for commit 7748d21c03b00066f48827977f8e6d904bd3c281. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/caiopizzol/cnpj-data-pipeline/pull/96?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: downloader.py [hunk 1]
```diff
@@ -214,6 +186,203 @@ def _download_and_extract(self, directory: str, filename: str) -> List[Path]:
 
         return extracted_files
 
+    def _download_zip(self, url: str, filename: str, zip_path: Path, log) -> None:
+        """Download a ZIP through a resumable .part file and validate it."""
+        part_path = zip_path.with_name(f"{zip_path.name}.part")
```
[the remark this answers is not in the document]
Comment (caiopizzol) [reply downloader.py]: Valid catch - fixed in e6991f9: the .part name now carries the source directory (month), so a stale partial from another month can never be resumed or 416-finalized against a different URL. Added test_stale_partial_from_other_month_is_not_resumed. (The final-zip cache-name collision under keep_files predates this PR and is left for a separate change.)

File: downloader.py [hunk 2]
```diff
@@ -214,6 +186,203 @@ def _download_and_extract(self, directory: str, filename: str) -> List[Path]:
...
+
+                logger.warning(f"Download attempt {attempt + 1} failed: {e}")
+                if attempt < self.config.retry_attempts - 1:
+                    time.sleep(self.config.retry_delay)
+                else:
+                    raise
+
+    def _download_zip_once(self, url: str, filename: str, zip_path: Path, part_path: Path) -> None:
+        offset = part_path.stat().st_size if part_path.exists() else 0
+        headers = {"Range": f"bytes={offset}-"} if offset else {}
```
[the remark this answers is not in the document]
Comment (caiopizzol) [reply downloader.py]: Fixed in e6991f9: all download requests now send Accept-Encoding: identity, so byte offsets always count identity bytes and ranged resumes can't be broken by a compressed representation.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,13 @@
 
 <!-- version list -->
 
+## Unreleased
```
[the remark this answers is not in the document]
Comment (caiopizzol) [reply CHANGELOG.md]: Agreed - the manual Unreleased entry would fight semantic-release's generated changelog. Reverted; the conventional commit message carries the release note.
